### PR TITLE
Support swig 4.0.0

### LINKED ISSUE #4727: force disable USE_SWIG_DIRECTORS in case of SWIG 4.0.0
swig 4.0 uses std::unexpected_handler when directors are enabled, which is removed in c++17  (https://en.cppreference.com/w/cpp/error/unexpected_handler). So if SWIG 4.0.0 is used we need to forcefully disable USE_SWIG_DIRECTORS cmake setting.

this will be fixed in 4.0.1 and later (see: https://github.com/swig/swig/blob/master/CHANGES.current#L37) so the check should be only for swig 4.0.0.

### PR DESCRIPTION
**Pull request by ghost**
Fixes #4727

### CODE REVIEW

File: src/shogun/util/factory.h [hunk 1]
```diff
@@ -276,14 +276,14 @@ namespace shogun
...
 	{
 		CDenseLabels* result = nullptr;
 		// try to interpret as all dense label types, from most restrictive to
 		// least restrictive
-		try_labels<CBinaryLabels>(result, labels);
-		try_labels<CMulticlassLabels>(result, labels);
-		try_labels<CRegressionLabels>(result, labels);
+		try_labels<CBinaryLabels>(result, labels_vector);
+		try_labels<CMulticlassLabels>(result, labels_vector);
+		try_labels<CRegressionLabels>(result, labels_vector);
```
Comment (ghost) [inline src/shogun/util/factory.h:286]: swig 4.0.0 had a bug when naming a function template argument with the same name as the function

File: CMakeLists.txt [hunk 2]
```diff
@@ -196,6 +196,12 @@ IF (${ANY_INTERFACE_ENABLED})
 	ENDIF()
 
 	FIND_PACKAGE(SWIG ${SWIG_VERSION_MINIMUM} REQUIRED)
+
+	IF(SWIG_VERSION EQUAL 4.0.0)
+		MESSAGE("Swig 4.0.0 doesn't support directors with c++17")
```
Comment (karlnapf) [inline CMakeLists.txt]: maybe we can add a "disabling" to the message?
Or, I would like an error even more as I don't like silent changes of settings....

Comment (karlnapf) [reply CMakeLists.txt]: This error was only around in exactly 4.0.0?

Comment (ghost) [reply CMakeLists.txt]: yes. it was introduced in 4.0.0, and will be fixed in 4.0.1.
do you mean something like
"Disabling swig directors: Not supported in swig 4.0.0"

Comment (karlnapf) [reply CMakeLists.txt]: yes.
Or "Due to a bug, Swig directors are not supported with Swig 4.0.0. Please disable or upgrade Swig"

Comment (ghost) [reply CMakeLists.txt]: ok, I changed it

**Review by karlnapf — APPROVED** [review]

**Pull request merged**